ocamltest/ocamltest_stdlib.ml

```diff
@@ -84,22 +84,40 @@ end
 module Sys = struct
   include Sys
 
-  let run_system_command prog args =
-    let command = Filename.quote_command prog args in
-    match Sys.command command with
-    | 0 -> ()
-    | _ as exitcode ->
-      Printf.eprintf "System command %s failed with status %d\n%!"
-        command exitcode;
-      exit 3
-
-  let mkdir dir =
-    if not (Sys.file_exists dir) then
-      run_system_command "mkdir" [dir]
+  let erase_file path =
+    try Sys.remove path
+    with Sys_error _ when (Sys.win32 || Sys.cygwin) &&
+                          Ocamltest_config.libunix <> None ->
+      (* Deal with read-only attribute on Windows. *)
+      Unix.chmod path 0o666;
+      Sys.remove path
+
+  let rm_rf path =
+    let rec erase_directory dir =
+      let remove_entry entry =
+        let path = Filename.concat dir entry in
+        if Sys.is_directory path then
+          erase_directory path
+        else
+          erase_file path
+      in
+        Array.iter remove_entry (Sys.readdir dir);
+        Sys.rmdir dir
+    in
+      if Sys.file_exists path then
+        if Sys.is_directory path then
+          erase_directory path
+        else
+          erase_file path
+      else ()
 
   let rec make_directory dir =
     if Sys.file_exists dir then ()
-    else (make_directory (Filename.dirname dir); mkdir dir)
+    else let () = make_directory (Filename.dirname dir) in
+         if not (Sys.file_exists dir) then
+           try Sys.mkdir dir 0o777
+           with _ -> exit 3
```

Comment:
Why `exit 3` ???  Why catch all errors here?

Reply:
I was also puzzled by `exit 3`; the reason is that this is current behaviour of `ocamltest`: https://github.com/ocaml/ocaml/blob/80326033cbedfe59c0664e3912f21017e968a1e5/ocamltest/ocamltest_stdlib.ml#L87-L94

For the catch-all, indeed it seems like it would be better to restrict to `Sys_error _`.

Reply:
@nojb is correct - and catching all errors instead of `Sys_error _` is a typo. An error message might be a good idea too.

Looking at the rest of ocamltest, we have exit code 1 for failed tests and invocation errors (bad command line, corrupt test etc.), exit code 2 for errors in a test itself (mainly used in `Tsl_semantics`) and then exit code 3 for errors in "external" commands, but that now is just this place.

Looking in `Main.tsl_block_of_file_safe`, perhaps this the handling could just be removed completely and allow that to display the `Sys_error` as it would with any other I/O problem?